File: src/decrypt/js/decrypt.js
#!/usr/bin/env node
// Usage: node decrypt.js "<ciphertext>" "<key>"
// If key omitted -> uses "default-key" to match the CLI fallback.

const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*_-?";
const NOISE_SET = "abcdefghijklmnopqrstuvwxyz";
const TERM = '~';
const BASE = 47;
const BLOCK_BYTES = 5;
const BLOCK_SYMBOLS = 8;
const CHECKSUM_PERIOD = 4;

function buildRev(alpha) {
	const m = new Map();
	for (let i = 0; i < alpha.length; i++) m.set(alpha[i], i);
	return m;
}
const REV_BASE = buildRev(ALPHABET);

function rotationForBlock(blockIndex) {
	return ((blockIndex * 13) + 11) % BASE;
}

function rotateAlpha(alpha, rot) {
	const n = alpha.length;
	rot = ((rot % n) + n) % n;
	return alpha.slice(rot) + alpha.slice(0, rot);
}

function base47DigitsTo5Bytes(digits) {
	if (digits.length !== BLOCK_SYMBOLS) throw new Error("expected 8 digits");
	// Use BigInt to be safe
	let val = 0n;
	for (const d of digits) {
		if (d < 0 || d >= BASE) throw new Error("digit out of range");
		val = val * BigInt(BASE) + BigInt(d);
	}
	const out = Buffer.alloc(BLOCK_BYTES);
	for (let i = BLOCK_BYTES - 1; i >= 0; i--) {
		out[i] = Number(val & 0xFFn);
		val >>= 8n;
	}
	return out;
}

function checksum47(blocksBytes) {
	let x = 0;
	for (const b of blocksBytes) {
		for (const byte of b) x ^= byte;
	}
	return x % BASE;
}

function isNoise(ch) {
	return NOISE_SET.includes(ch);
}

function decodeMosaic(s) {
	let i = 0;
	const n = s.length;
	const outParts = [];
	let blockIndex = 0;
	let csWindows = [];
	const revBase = REV_BASE;

	while (i < n) {
		// skip whitespace
		while (i < n && /\s/.test(s[i])) i++;
		if (i >= n) break;

		// trailer detection: "~~" + one base47 digit
		if (i + 2 < n && s[i] === TERM && s[i + 1] === TERM) {
			const padChar = s[i + 2];
			if (!revBase.has(padChar)) throw new Error(`Invalid trailer pad digit: ${padChar}`);
			const padCount = revBase.get(padChar);
			if (padCount < 0 || padCount >= BLOCK_BYTES) throw new Error(`Invalid pad_count: ${padCount}`);
			// apply pad removal
			let assembled = Buffer.concat(outParts);
			if (padCount) {
				if (assembled.length < padCount) throw new Error("Pad mismatch (output too small to trim)");
				assembled = assembled.slice(0, assembled.length - padCount);
			}
			i += 3;
			if (i !== n) throw new Error("Extra data after trailer");
			return assembled;
		}

		const rot = rotationForBlock(blockIndex);
		const rotated = rotateAlpha(ALPHABET, rot);
		const revRot = buildRev(rotated);

		// read 8 digits skipping noise
		const digits = [];
		for (let k = 0; k < BLOCK_SYMBOLS; k++) {
			while (i < n && isNoise(s[i])) i++;
			if (i >= n) throw new Error("Unexpected end of input while reading digits");
			const c = s[i];
			if (c === TERM) throw new Error("Unexpected terminator while expecting digit");
			if (!revRot.has(c)) throw new Error(`Invalid digit character for rotated alphabet: ${c}`);
			digits.push(revRot.get(c));
			i++;
		}

		// skip noise then expect terminator
		while (i < n && isNoise(s[i])) i++;
		if (i >= n || s[i] !== TERM) throw new Error("Missing block terminator after digits");
		i++; // consume terminator

		const block5 = base47DigitsTo5Bytes(digits);
		outParts.push(block5);

		csWindows.push(block5);
		if (csWindows.length > CHECKSUM_PERIOD) csWindows = csWindows.slice(-CHECKSUM_PERIOD);

		blockIndex++;

		if (csWindows.length === CHECKSUM_PERIOD) {
			while (i < n && isNoise(s[i])) i++;
			if (i >= n) throw new Error("Missing checksum character after block group");
			const chkChar = s[i++];
			if (!revBase.has(chkChar)) throw new Error(`Invalid checksum char: ${chkChar}`);
			const got = revBase.get(chkChar);
			const expect = checksum47(csWindows);
			if (got !== expect) throw new Error(`Checksum mismatch: got ${got} expect ${expect}`);
			csWindows = [];
		}
	}

	throw new Error("No trailer found; malformed input");
}

function xorWithKey(rawBuf, keyBuf) {
	if (!keyBuf || keyBuf.length === 0) return rawBuf;
	const out = Buffer.alloc(rawBuf.length);
	for (let i = 0; i < rawBuf.length; i++) out[i] = rawBuf[i] ^ keyBuf[i % keyBuf.length];
	return out;
}

/* --- main --- */
function main() {
	const argv = process.argv.slice(2);
	if (argv.length < 1) {
		console.error("Usage: node mosaic_decode.js '<ciphertext>' [key]");
		process.exit(1);
	}
	const s = argv[0];
	const key = argv.length >= 2 ? argv[1] : "default-key";

	let raw;
	try {
		raw = decodeMosaic(s);
	} catch (e) {
		console.error("Decoding error:", e.message || e);
		process.exit(2);
	}

	const plain = xorWithKey(raw, Buffer.from(key, "utf8"));
	console.log("Decoded bytes (hex):", plain.toString("hex"));

	try {
		const text = plain.toString("utf8");
		// verify valid utf8 round-trip: Buffer -> string -> Buffer
		if (Buffer.from(text, "utf8").equals(plain)) {
			console.log("Decoded text (utf-8):", text);
		} else {
			throw new Error("not valid utf8");
		}
	} catch (_) {
		// escaped printable fallback
		let out = "";
		for (const b of plain) {
			if (b >= 32 && b <= 126) out += String.fromCharCode(b);
			else out += `\\x${b.toString(16).padStart(2, "0")}`;
		}
		console.log("Decoded text:", out);
	}
}

if (require.main === module) main();

File: src/decrypt/python/decrypt.py
#!/usr/bin/env python3

import sys

ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*_-?"
NOISE_SET = "abcdefghijklmnopqrstuvwxyz"
TERM = '~'
BASE = 47
BLOCK_BYTES = 5
BLOCK_SYMBOLS = 8
CHECKSUM_PERIOD = 4

def build_rev(alpha):
    return {ch: i for i, ch in enumerate(alpha)}

REV_BASE = build_rev(ALPHABET)

def rotation_for_block(block_index):
    return ((block_index * 13) + 11) % BASE

def rotate_alpha(alpha, rot):
    n = len(alpha)
    rot = rot % n
    return ''.join(alpha[(i + rot) % n] for i in range(n))

def base47_digits_to_5bytes(digits):
    val = 0
    for d in digits:
        val = val * BASE + d
    out = bytearray(BLOCK_BYTES)
    for i in range(BLOCK_BYTES - 1, -1, -1):
        out[i] = val & 0xFF
        val >>= 8
    return bytes(out)

def checksum47(blocks_bytes):
    x = 0
    for b in blocks_bytes:
        for byte in b:
            x ^= byte
    return x % BASE

def decode_mosaic(s):
    i = 0
    n = len(s)
    out = bytearray()
    block_index = 0
    cs_windows = []
    rev_base = REV_BASE

    while i < n:
        # skip whitespace
        while i < n and s[i].isspace():
            i += 1
        if i >= n:
            break

        # trailer: "~~" + paddigit
        if i + 2 < n and s[i] == TERM and s[i+1] == TERM:
            pad_char = s[i+2]
            if pad_char not in rev_base:
                raise ValueError("Invalid trailer pad digit: %r" % pad_char)
            pad_count = rev_base[pad_char]
            if pad_count < 0 or pad_count >= BLOCK_BYTES:
                raise ValueError("Invalid pad_count: %d" % pad_count)
            if pad_count:
                if len(out) < pad_count:
                    raise ValueError("Pad mismatch (output too small to trim)")
                out = out[:-pad_count]
            i += 3
            if i != n:
                raise ValueError("Extra data after trailer")
            return bytes(out)

        rot = rotation_for_block(block_index)
        rotated = rotate_alpha(ALPHABET, rot)
        rev_rot = build_rev(rotated)

        digits = []
        for k in range(BLOCK_SYMBOLS):
            while i < n and s[i] in NOISE_SET:
                i += 1
            if i >= n:
                raise ValueError("Unexpected end while reading digits")
            c = s[i]
            if c == TERM:
                raise ValueError("Unexpected terminator while expecting digit")
            if c not in rev_rot:
                raise ValueError("Invalid digit: %r" % c)
            digits.append(rev_rot[c])
            i += 1

        while i < n and s[i] in NOISE_SET:
            i += 1
        if i >= n or s[i] != TERM:
            raise ValueError("Missing block terminator after digits")
        i += 1

        block5 = base47_digits_to_5bytes(digits)
        out.extend(block5)

        cs_windows.append(block5)
        if len(cs_windows) > CHECKSUM_PERIOD:
            cs_windows = cs_windows[-CHECKSUM_PERIOD:]

        block_index += 1

        if len(cs_windows) == CHECKSUM_PERIOD:
            while i < n and s[i] in NOISE_SET:
                i += 1
            if i >= n:
                raise ValueError("Missing checksum character after block group")
            chk_char = s[i]; i += 1
            if chk_char not in rev_base:
                raise ValueError("Invalid checksum char: %r" % chk_char)
            got = rev_base[chk_char]
            expect = checksum47(cs_windows)
            if got != expect:
                raise ValueError("Checksum mismatch: got %d expect %d" % (got, expect))
            cs_windows = []

    raise ValueError("No trailer found; malformed input")

def xor_with_key(raw, key_bytes):
    if not key_bytes:
        return raw
    out = bytearray(len(raw))
    klen = len(key_bytes)
    for i, b in enumerate(raw):
        out[i] = b ^ key_bytes[i % klen]
    return bytes(out)

def main():
    if len(sys.argv) < 2:
        print("Usage: decrypt.py '<ciphertext>' [key]")
        return 1
    s = sys.argv[1]
    # if key omitted, use same fallback as C CLI
    key = sys.argv[2] if len(sys.argv) >= 3 else "default-key"

    try:
        raw = decode_mosaic(s)
    except Exception as e:
        print("Decoding error:", e)
        return 2

    raw = xor_with_key(raw, key.encode('utf-8'))

    print("Decoded bytes (hex):", raw.hex())
    try:
        text = raw.decode('utf-8')
        print("Decoded text (utf-8):", text)
    except Exception:
        # printable fallback
        printable = ''.join((chr(b) if 32 <= b <= 126 else "\\x%02x" % b) for b in raw)
        print("Decoded text:", printable)
    return 0

if __name__ == "__main__":
    sys.exit(main())
